### ISSUE TITLE
ImageUpdateAutomation creates multiple (incorrect) commits for 1 update

### ISSUE THREAD

**Issue by rmoesbergen** [fluxcd/image-automation-controller#482]
### Describe the bug

I have setup an ImageUpdateAutomation to update docker image tags in a git repo containing a bunch of yaml files. The files contain update tags, like this:

imageTag: v5.3.0 # {"$imagepolicy": "flux-system:backend-production:tag"}

The ImageUpdateAutomation:
```
spec:
  git:
    checkout:
      ref:
        branch: candidates
    commit:
      author:
        email: bla@...
        name: github-ci
      messageTemplate: '{{range .Updated.Images}}{{println .}}{{end}}'
    push:
      branch: candidates
  interval: 10m0s
  sourceRef:
    kind: GitRepository
    name: config
  update:
    path: ./k8s/flux
    strategy: Setters
```

The image policy looks as follows:
```
spec:
  imageRepositoryRef:
    name: backend
  policy:
    semver:
      range: '>v1.0.0'
```

This all works as expected, except that when an updated image tag is pushed, the image update controller creates a bunch of commits switching back and forth between the current and the new tag, like this:

```
commit ded9165a3613e4c91afba1227e6892a26d6789a5
Date:   Fri Feb 3 23:31:24 2023 +0000

    eu.gcr.io/company-cloud/company-backend:v5.3.1

diff --git a/k8s/flux/production/shared/helm-backend.yaml b/k8s/flux/production/shared/helm-backend.yaml
index 86239bde..1850e6c5 100644
--- a/k8s/flux/production/shared/helm-backend.yaml
+++ b/k8s/flux/production/shared/helm-backend.yaml
@@ -19,7 +19,7 @@ spec:
     remediation:
       retries: -1
   values:
-    imageTag: v5.3.0 # {"$imagepolicy": "flux-system:backend-production:tag"}
+    imageTag: v5.3.1 # {"$imagepolicy": "flux-system:backend-production:tag"}
   valuesFrom:
   - kind: ConfigMap
     name: helm-values

commit 29aa71f35eb74940ac963f9cbd0e9b5616312250
Date:   Fri Feb 3 23:31:16 2023 +0000

    eu.gcr.io/company-cloud/company-backend:v5.3.0

diff --git a/k8s/flux/production/shared/helm-backend.yaml b/k8s/flux/production/shared/helm-backend.yaml
index 1850e6c5..86239bde 100644
--- a/k8s/flux/production/shared/helm-backend.yaml
+++ b/k8s/flux/production/shared/helm-backend.yaml
@@ -19,7 +19,7 @@ spec:
     remediation:
       retries: -1
   values:
-    imageTag: v5.3.1 # {"$imagepolicy": "flux-system:backend-production:tag"}
+    imageTag: v5.3.0 # {"$imagepolicy": "flux-system:backend-production:tag"}
   valuesFrom:
   - kind: ConfigMap
     name: helm-values

commit 8fc27bdef09cda5339fa349f8b76cb037bcf7d35
Date:   Fri Feb 3 23:31:07 2023 +0000

    eu.gcr.io/company-cloud/company-backend:v5.3.1

diff --git a/k8s/flux/production/shared/helm-backend.yaml b/k8s/flux/production/shared/helm-backend.yaml
index 86239bde..1850e6c5 100644
--- a/k8s/flux/production/shared/helm-backend.yaml
+++ b/k8s/flux/production/shared/helm-backend.yaml
@@ -19,7 +19,7 @@ spec:
     remediation:
       retries: -1
   values:
-    imageTag: v5.3.0 # {"$imagepolicy": "flux-system:backend-production:tag"}
+    imageTag: v5.3.1 # {"$imagepolicy": "flux-system:backend-production:tag"}
   valuesFrom:
   - kind: ConfigMap
     name: helm-values
```

So in this case the new tag was v5.3.1, which was updated in the first commit, but then there's a second commit with v5.3.0 (previous version), and then another v5.3.1 commit. I would expect there to be just 1 commit with the new version tag.

### Steps to reproduce

1. Set up an image repository
2. Create an ImageUpdateAutomation and ImagePolicy as specified above
3. Push a new image to the repo with a new tag
4. Multiple commits created instead of 1

### Expected behavior

I would expect only 1 commit with the new version.

### OS / Distro

N/A

### Flux version

flux: v0.36.0

### Flux check

► checking prerequisites
✗ flux 0.36.0 <0.39.0 (new version is available, please upgrade)
✔ Kubernetes 1.24.8-gke.2000 >=1.20.6-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.26.0
✔ image-automation-controller: deployment ready
► ghcr.io/fluxcd/image-automation-controller:v0.26.1
✔ image-reflector-controller: deployment ready
► ghcr.io/fluxcd/image-reflector-controller:v0.22.1
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v0.30.0
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v0.28.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v0.31.0
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta1
✔ buckets.source.toolkit.fluxcd.io/v1beta1
✔ gitrepositories.source.toolkit.fluxcd.io/v1beta1
✔ helmcharts.source.toolkit.fluxcd.io/v1beta1
✔ helmreleases.helm.toolkit.fluxcd.io/v2beta1
✔ helmrepositories.source.toolkit.fluxcd.io/v1beta1
✔ imagepolicies.image.toolkit.fluxcd.io/v1beta1
✔ imagerepositories.image.toolkit.fluxcd.io/v1beta1
✔ imageupdateautomations.image.toolkit.fluxcd.io/v1beta1
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1beta2
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta1
✔ receivers.notification.toolkit.fluxcd.io/v1beta1
✔ all checks passed

### Git provider

GitHub

### Container Registry provider

Google Container Registry (GCR)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by makkes** [comment]
I suppose to reproduce the issue it should suffice to create an `ImageRepository` and an `ImagePolicy` and watch the events.

**Comment by stefanprodan** [comment]
I have a potential fix in #484, please test it using  `ghcr.io/fluxcd/image-automation-controller:rc-4949c9ff`.

**Comment by rmoesbergen** [comment]
I tried with the updated automation-controller, but that seems to make the problem even worse: I now get 5 commits for 1 image update, of which 3 are logged in events:

```
0s          Normal    info                         imageupdateautomation/backend              Committed and pushed change 9209b3135b4f9b39a69b5d6959c547e76ab4ae5b to candidates...
0s          Normal    GitOperationSucceeded        gitrepository/helmcharts                   no changes since last reconcilation: observed revision 'main/8fc29bb0df884c25de24b8a5447e6652aac90fbd'
0s          Normal    NewArtifact                  gitrepository/config                       stored artifact for commit 'eu.gcr.io/cloud/admin:main-d4871...'
0s          Normal    GarbageCollectionSucceeded   gitrepository/config                       garbage collected 1 artifacts
0s          Normal    GitOperationSucceeded        gitrepository/helmcharts                   no changes since last reconcilation: observed revision 'main/8fc29bb0df884c25de24b8a5447e6652aac90fbd'
0s          Normal    NewArtifact                  gitrepository/config                       stored artifact for commit 'eu.gcr.io/cloud/admin:main-d4871...'
0s          Normal    info                         imageupdateautomation/backend              Committed and pushed change 5897d10178f02a282e3704f7e25575f31082965e to candidates...
0s          Normal    GitOperationSucceeded        gitrepository/helmcharts                   no changes since last reconcilation: observed revision 'main/8fc29bb0df884c25de24b8a5447e6652aac90fbd'
0s          Normal    NewArtifact                  gitrepository/config                       stored artifact for commit 'eu.gcr.io/cloud/admin:main-d4871...'
0s          Normal    GarbageCollectionSucceeded   gitrepository/config                       garbage collected 1 artifacts
0s          Normal    GitOperationSucceeded        gitrepository/helmcharts                   no changes since last reconcilation: observed revision 'main/8fc29bb0df884c25de24b8a5447e6652aac90fbd'
0s          Normal    info                         imageupdateautomation/backend              Committed and pushed change 3798426cbf858b5f4a1902d141124c0f74b00e1c to candidates...
0s          Normal    GarbageCollectionSucceeded   gitrepository/config                       garbage collected 1 artifacts
0s          Normal    GitOperationSucceeded        gitrepository/helmcharts                   no changes since last reconcilation: observed revision 'main/8fc29bb0df884c25de24b8a5447e6652aac90fbd'
0s          Normal    GitOperationSucceeded        gitrepository/config                       no changes since last reconcilation: observed revision 'candidates/3798426cbf858b5f4a1902d141124c0f74b00e1c'
```
Which looks like this in git:
```
commit 3798426cbf858b5f4a1902d141124c0f74b00e1c (HEAD -> candidates, origin/candidates, origin/HEAD)
Date:   Thu Feb 9 10:51:58 2023 +0000

    eu.gcr.io/cloud/admin:main-d4871a6-1675939762

commit 3209d2302744626d15c213b4173e32f1e93f3e37
Date:   Thu Feb 9 10:51:55 2023 +0000

    eu.gcr.io/cloud/admin:main-d4871a6-1675850003

commit 5897d10178f02a282e3704f7e25575f31082965e
Date:   Thu Feb 9 10:51:48 2023 +0000

    eu.gcr.io/cloud/admin:main-d4871a6-1675939762

commit 5d9e18d8ee207eaafd6b49214244e7ec6e1440f1
Date:   Thu Feb 9 10:51:43 2023 +0000

    eu.gcr.io/cloud/admin:main-d4871a6-1675850003

commit 9209b3135b4f9b39a69b5d6959c547e76ab4ae5b
Date:   Thu Feb 9 10:51:36 2023 +0000

    eu.gcr.io/cloud/admin:main-d4871a6-1675939762

```

**Comment by stefanprodan** [comment]
Hmm so every couple of seconds the policy reports a different tag. Can you please post here the policy using `kubectl get imagepolicy --show-managed-fields -oyaml`

**Comment by rmoesbergen** [comment]
I have multiple imagepolicies, but this is the one referenced above (for 'admin'):
```
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImagePolicy
metadata:
  annotations:
    meta.helm.sh/release-name: flux
    meta.helm.sh/release-namespace: flux-system
  creationTimestamp: "2022-01-26T11:11:28Z"
  finalizers:
  - finalizers.fluxcd.io
  generation: 1
  labels:
    app.kubernetes.io/managed-by: Helm
  managedFields:
  - apiVersion: image.toolkit.fluxcd.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:meta.helm.sh/release-name: {}
          f:meta.helm.sh/release-namespace: {}
        f:labels:
          .: {}
          f:app.kubernetes.io/managed-by: {}
      f:spec:
        .: {}
        f:filterTags:
          .: {}
          f:extract: {}
          f:pattern: {}
        f:imageRepositoryRef:
          .: {}
          f:name: {}
        f:policy:
          .: {}
          f:numerical:
            .: {}
            f:order: {}
    manager: terraform-provider-helm_v2.4.1_x5
    operation: Update
    time: "2022-01-26T11:11:28Z"
  - apiVersion: image.toolkit.fluxcd.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:conditions: {}
        f:latestImage: {}
    manager: image-reflector-controller
    operation: Update
    subresource: status
    time: "2022-11-14T11:01:45Z"
  - apiVersion: image.toolkit.fluxcd.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .: {}
          v:"finalizers.fluxcd.io": {}
      f:status:
        f:observedGeneration: {}
    manager: image-reflector-controller
    operation: Update
    time: "2022-11-17T09:21:26Z"
  name: admin-staging
  namespace: flux-system
  resourceVersion: "748935489"
  uid: ac9b2578-5701-4f37-a288-36b13f9d4990
spec:
  filterTags:
    extract: $ts
    pattern: ^main-[a-fA-F0-9]+-(?P<ts>.*)
  imageRepositoryRef:
    name: admin
  policy:
    numerical:
      order: asc
status:
  conditions:
  - lastTransitionTime: "2022-11-14T11:01:45Z"
    message: 'Latest image tag for ''eu.gcr.io/cloud/admin'' resolved
      to: main-d4871a6-1675939762'
    reason: ReconciliationSucceeded
    status: "True"
    type: Ready
  latestImage: eu.gcr.io/cloud/admin:main-d4871a6-1675939762
  observedGeneration: 1
```

**Comment by stefanprodan** [comment]
Can you also post here `kubectl describe imagepolicy`, I'm interested in the events section.

**Comment by rmoesbergen** [comment]
Sure:
```
Events:
  Type    Reason  Age                  From                        Message
  ----    ------  ----                 ----                        -------
  Normal  info    51m (x61 over 112m)  image-reflector-controller  Latest image tag for 'eu.gcr.io/deepdesk-cloud/deepdesk-admin' resolved to: main-d4871a6-1675850003
  Normal  info    96s (x47 over 47m)   image-reflector-controller  Latest image tag for 'eu.gcr.io/deepdesk-cloud/deepdesk-admin' resolved to: main-d4871a6-1675939762
```

**Comment by stefanprodan** [comment]
Ok so the issue is not in image-automation-controller, the policy flips the tag from what it sees in the registry and every time it detects a different tag, image-automation-controller does a new commit.

**Comment by stefanprodan** [comment]
After an image push, do you trigger an ImageRepository reconciliation with a webhook or via the flux CLI, or do you wait for the interval? Can you also post here `kubectl get imageregistry --show-managed-fields -oyaml` and `kubectl describe imageregistry` events.

**Comment by rmoesbergen** [comment]
The ImageRepository only uses the scanning interval, GCR does not have the ability to push to a webhook on image push. The interval is now set to 1 minute. The git commits happen within a minute, so that's strange.

```
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImageRepository
metadata:
  annotations:
    meta.helm.sh/release-name: flux
    meta.helm.sh/release-namespace: flux-system
  creationTimestamp: "2022-01-26T11:11:28Z"
  finalizers:
  - finalizers.fluxcd.io
  generation: 2
  labels:
    app.kubernetes.io/managed-by: Helm
  managedFields:
  - apiVersion: image.toolkit.fluxcd.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:meta.helm.sh/release-name: {}
          f:meta.helm.sh/release-namespace: {}
        f:labels:
          .: {}
          f:app.kubernetes.io/managed-by: {}
      f:spec:
        .: {}
        f:image: {}
        f:interval: {}
    manager: terraform-provider-helm_v2.4.1_x5
    operation: Update
    time: "2022-01-26T11:11:28Z"
  - apiVersion: image.toolkit.fluxcd.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:conditions: {}
        f:lastScanResult:
          f:scanTime: {}
          f:tagCount: {}
    manager: image-reflector-controller
    operation: Update
    subresource: status
    time: "2022-06-08T12:00:08Z"
  - apiVersion: image.toolkit.fluxcd.io/v1beta1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .: {}
          v:"finalizers.fluxcd.io": {}
      f:status:
        f:canonicalImageName: {}
        f:lastScanResult: {}
        f:observedGeneration: {}
    manager: image-reflector-controller
    operation: Update
    time: "2022-11-17T09:21:26Z"
  name: admin
  namespace: flux-system
  resourceVersion: "749017363"
  uid: 22fce7d0-e494-4bab-9768-18a69d30e81b
spec:
  image: eu.gcr.io/cloud/admin
  interval: 1m0s
status:
  canonicalImageName: eu.gcr.io/cloud/admin
  conditions:
  - lastTransitionTime: "2022-12-05T12:38:47Z"
    message: successful scan, found 85 tags
    reason: ReconciliationSucceeded
    status: "True"
    type: Ready
  lastScanResult:
    scanTime: "2023-02-09T11:45:48Z"
    tagCount: 85
  observedGeneration: 2
```

Events is empty, k8s cleaned them up :(

**Comment by rmoesbergen** [comment]
Thinking about webhooks: the GitRepository _does_ have a webhook towards flux, which is also the repo that the image automation controller is pushing updates to, maybe the 2 are in eachothers way? So this is what happens:

- Image automation sees new tag
- Push new commit to 'config' repo
- Webhook called for the 'config' repo -> triggers the flux source controller to repackage the repo

Maybe then the image automation controllers gets the 'config' repo update with its own change, and gets confused somehow? Just 'thinking' out loud here.

**Comment by rmoesbergen** [comment]
I think I've found the problem... and it's not a bug in flux. We have multiple imagepolicies and one ImageAutomation in our staging k8s cluster. These policies monitor staging images and push to a central configuration repo. So far so good. However, we also have several other (production) clusters that also have imagepolicies and one imageAutomation in place. As it turns out, the staging imagepolicies are also active on the production clusters 🤦🏻‍♂️ So whenever an image is pushed, all clusters start fighting in the config repo until all clusters have the new imagetag (they all have a polling interval of 1 minute). This also explains the git commits that were not logged: they are, but just on a different cluster... So obvious in hindsight 😞 

To answer your question about helm: I use `flux install --export` to generate a helm chart and then use terraform to deploy that chart everywhere. I saw there's a terraform provider now, but that didn't exist when we started using flux.

Anyhow, I think it's safe to say this "bug" can be closed. Thanks for your patience and sorry to waste your time.
@madcla Does any of the above maybe also explain your situation (Multiple ImageAutomations, and/or multiple clusters)?

**Comment by madcla** [comment]
@rmoesbergen Oh yes. Obviously! :roll_eyes: Now that you mention it we actually also have some dynamic clusters spinning up and some of them staging the application as well. Makes sense that they then also commit to the repo. Thanks for the hint!

**Comment by ale900522** [comment]
> Ok so the issue is not in image-automation-controller, the policy flips the tag from what it sees in the registry and every time it detects a different tag, image-automation-controller does a new commit.

I have this config and I receive this error: 
`Events:                                                                                                                                                                                                           │
│   Type     Reason  Age                 From                         Message                                                                                                                                       │
│   ----     ------  ----                ----                         -------                                                                                                                                       │
│   Warning  error   64s (x11 over 13m)  image-automation-controller  authorization failed`

I have a imagerepo:
`---
apiVersion: image.toolkit.fluxcd.io/v1beta2
kind: ImageRepository
metadata:
  name: api-repo-gitlab
  namespace: flux-system
spec:
  exclusionList:
  - ^.*\.sig$
  image: registry.gitlab.com/api-repo-gitlab
  interval: 5m0s
  provider: generic
  secretRef:
    name: regcred`
The result is:
` Last Scan Result:                                                                                                                                                                                               
     Latest Tags:                                                                                                                                                                                                  
       latest                                                                                                                                                                                                      
       0.0.2                                                                                                                                                                                                       
       0.0.1                                                                                                                                                                                                       
     Scan Time:  2024-01-04T12:46:24Z                                                                                                                                                                              
     Tag Count:  3 ` Perfect for me.
My imagepolicy:
`---
apiVersion: image.toolkit.fluxcd.io/v1beta2
kind: ImagePolicy
metadata:
  name: api-repo-gitlab-policy
  namespace: flux-system
spec:
  imageRepositoryRef:
    name: api-repo-gitlab
  policy:
    semver:
      range: '>=0.0.1' `
Basic config but is ok, the result is:
`Events:                                                                                                                                                                                                           
   Type    Reason     Age                  From                        Message                                                                                                                                    
   ----    ------     ----                 ----                        -------                                                                                                                                     
   Normal  Succeeded  84s (x78 over 131m)  image-reflector-controller  Latest image tag for 'registry.gitlab.com/api-repo-gitlab' resolved to 0.0.2 `
Result is ok

My image update is:
`---
apiVersion: image.toolkit.fluxcd.io/v1beta1
kind: ImageUpdateAutomation
metadata:
  name: api-repo-gitlab  
  namespace: flux-system
spec:
  git:
    checkout:
      ref:
        branch: main
    commit:
      author:
        email: admin@example.com
        name: admin
      messageTemplate: |-
        {{ range .Updated.Images -}}
        [demo] Automated image update **{{ $.AutomationObject }}** to **{{ .Identifier }}**
        {{ end -}}
        Automation name: {{ .AutomationObject }}

        Files:
        {{ range $filename, $_ := .Updated.Files -}}
        - {{ $filename }}
        {{ end -}}

        Objects:
        {{ range $resource, $_ := .Updated.Objects -}}
        - {{ $resource.Kind }} {{ $resource.Name }}
        {{ end -}}

        Images:
        {{ range .Updated.Images -}}
        - {{.}}
        {{ end -}}
  interval: 1m0s
  sourceRef:
    kind: GitRepository
    name: api-repo-gitlab  #this GitRepositorie exit in flux-system namespace
    namespace: flux-system
  update:
    strategy: Setters`

I run fluxcd boostrap with access-token and this is my boostrap
`flux bootstrap gitlab \
           --components-extra=image-reflector-controller,image-automation-controller \
           --read-write-key \
           --deploy-token-auth \
           --owner=learn \
           --repository=gitops-fluxcd \
           --branch=master \
           --path=k8s-cluster`
Thank you !!!
